### PR TITLE
7903618: JMH: Add power events to perf profilers

### PR DESCRIPTION
**Pull request by schrieveslaach**
The Linux tool perf provides collecting energy consumption via Intel's RAPL API. Today's climate crisis does not need to have tools to be as performant as possible it also requires software to be as energy efficient as possible. Thus measuring the energy consumption of processes is a crucial tool.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Committer](https://openjdk.org/bylaws#committer))

### Issue
 * [CODETOOLS-7903618](https://bugs.openjdk.org/browse/CODETOOLS-7903618): JMH: Add power events to perf profilers (**Enhancement** - P4)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jmh.git pull/126/head:pull/126` \
`$ git checkout pull/126`

Update a local copy of the PR: \
`$ git checkout pull/126` \
`$ git pull https://git.openjdk.org/jmh.git pull/126/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 126`

View PR using the GUI difftool: \
`$ git pr show -t 126`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jmh/pull/126.diff">https://git.openjdk.org/jmh/pull/126.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jmh/pull/126#issuecomment-1901185613)

### CODE REVIEW

**Comment by bourgesl** [comment]
Nice initiative, although repl does not report the idle power consumption...

File: jmh-core/src/main/java/org/openjdk/jmh/profile/LinuxPerfProfiler.java [hunk 1]
```diff
@@ -224,20 +247,42 @@ protected Aggregator<PerfResult> getIterationAggregator() {
...
-                return String.format("%s IPC, %s CPI",
+                StringBuilder str = new StringBuilder();
+
+                str.append(String.format("%s IPC, %s CPI",
                         ScoreFormatter.format(1.0 * instructions / cycles),
-                        ScoreFormatter.format(1.0 * cycles / instructions));
+                        ScoreFormatter.format(1.0 * cycles / instructions)));
+
+
+                if(!Double.isNaN(joulesCpu)) {
```
Comment (shipilev) [inline jmh-core/src/main/java/org/openjdk/jmh/profile/LinuxPerfProfiler.java:278]:
```suggestion
                if (!Double.isNaN(joulesCpu)) {
```

File: jmh-core/src/main/java/org/openjdk/jmh/profile/LinuxPerfProfiler.java [hunk 2]
```diff
@@ -224,20 +247,42 @@ protected Aggregator<PerfResult> getIterationAggregator() {
...
+                    str.append(String.format(", %s CPU Joules", ScoreFormatter.format(joulesCpu)));
+                }
+                if(!Double.isNaN(joulesRam)) {
```
Comment (shipilev) [inline jmh-core/src/main/java/org/openjdk/jmh/profile/LinuxPerfProfiler.java:281]:
```suggestion
                if (!Double.isNaN(joulesRam)) {
```

**Review by shipilev — COMMENTED** [review]
Apologies for not looking into this more promptly. This looks interesting, thanks! 

Are the new events always printed by `perf stat`, or should we request them specifically? Are you planning to amend `perfnorm`/`perfasm` profilers as well, or just supplying the event name there works already?

**Comment by shipilev** [comment]
Could you also go to https://github.com/schrieveslaach/jmh/actions -- and enable testing workflows for your fork? This would test things. Thanks!

**Comment by shipilev** [comment]
@schrieveslaach Do you still have interest in completing this? If not, I can take over.

**Pull request closed without merging**